pkg/kubelet/cm/dra/manager.go

```diff
@@ -83,10 +72,9 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				claimName := resourceclaim.Name(pod, &pod.Spec.ResourceClaims[i])
 				klog.V(3).InfoS("Processing resource", "claim", claimName, "pod", pod.Name)
 
+				// Resource is already prepared, add pod UID to it
 				if claimInfo := m.cache.get(claimName, pod.Namespace); claimInfo != nil {
-					// resource is already prepared, add pod UID to it
 					claimInfo.addPodReference(pod.UID)
-
```

Comment:
We should consider checkpointing here and in the similar places.
Actually, if we decide to checkpoint on every change we may want to return back to checkpointing in the claimInfo methods (`add`, `delete`, `addPodReference` etc).

Reply:
I think the motivation for also checkout after NodePrepareResource to avoid calling the grpc again if we didn't checkpoint it at the end of the PrepareResources.  for example one DRA plugin succeed (checkpoint it) second DRA plugin failed. 

And the reason for UnNodePrepareResource is to be similar to PrepareResources flow. 

now I am rethinking that anyway we can't cover all the corner cases see [1] with this logic so maybe we will do checkpoint at just the end of the PrepareResources/UnPrepareResources  for this phase. 

[1] - 
the corner case we try to provide solution was:
1. Pod created
2. Pod passed to kubelet
3. PrepareResources call is made in the kubelet
4. Some number of underlying gRPC calls to plugins’ PrepareResources succeeds
5. kubelet is taken offline before checkpoint happens
6. PrepareResources call to the kubelet times out
7. Pod enters failed state because it was never able to be started on the node
8. Pod gets deleted
9. Plugins have state about a now deleted pod without the ability to ever recover from that

Reply:
I don't like the idea of the checkpointing being implicit in the various `add()` and `delete()` calls.

It would be fine to move this call up and call it more frequently, but I feel strongly about not embedding it in the calls.

Reply:
I'm ok with both ways of doing it. In fact we seem to do it on any cache change anyway.